hepcrawl/spiders/base_spider.py

```diff
@@ -0,0 +1,229 @@
+            "jpg" not in identifier.lower()
+        ]
+        relations = [
+            s for s in " ".join(node.xpath(".//dc:relation/text()").extract()).split() if "http" in s and
+            "jpg" not in s.lower()
+        ]
+        links = node.xpath(".//base_dc:link/text()").extract()
+        urls_in_record = []
+        for url in identifiers + relations + links:
+            if url.startswith('<'):
+                url = url[1:]
+            if url.endswith('>'):
+                url = url[:-1]
+            if not url.startswith("http://") and not url.startswith("https://"):
+                url = "http://{0}".format(url)
+            if url not in urls_in_record:
+                urls_in_record.append(url)
+        return urls_in_record
+
+    def find_direct_links(self, urls_in_record):
+        """Determine if the XML file has a direct link."""
+        direct_link = []
+        for link in urls_in_record:
+            if "pdf" in get_mime_type(link) and "jpg" not in link.lower():
+                direct_link.append(link)
+        if direct_link:
+            self.logger.info("Found direct link(s): %s", direct_link)
+        else:
+            self.logger.info("Didn't find direct link to PDF")
+
+        return direct_link
+
+    def parse_node(self, response, node):
+        """Iterate through all the record nodes in the XML.
+
+        With each node it checks if direct link exists, and sends
+        a request to appropriate function to parse the XML.
+        """
+        urls_in_record = self.get_urls_in_record(node)
+        direct_link = self.find_direct_links(urls_in_record)
+
+        if not direct_link and urls_in_record:
+            # Probably all links lead to same place, so take first
+            link = urls_in_record[0]
+            request = Request(link, callback=self.scrape_for_pdf)
+            request.meta["urls"] = urls_in_record
+            request.meta["node"] = node
+            return request
+        else:
+            response.meta["direct_link"] = direct_link
+            response.meta["urls"] = urls_in_record
+            response.meta["node"] = node
+            return self.build_item(response)
+
+    def build_item(self, response):
+        """Build the final record."""
+        node = response.meta["node"]
+        record = HEPLoader(item=HEPRecord(), selector=node, response=response)
+        record.add_value('files', response.meta.get("direct_link"))
+        record.add_value('urls', response.meta.get("urls"))
+        record.add_xpath('abstract', './/dc:description/text()')
+        record.add_xpath('title', './/dc:title/text()')
+        record.add_xpath('date_published', './/dc:date/text()')
+        record.add_xpath('source', './/base_dc:collname/text()')
+        # Should this be able to scrape all kinds of publications?
+        # Now does only theses:
+        record.add_value('thesis', {'degree_type': 'PhD'})
```

Comment:
Actually, will we always harvest thesis from base? If not, this information should only be added conditionally. @annetteholtkamp